ABOUT DATASET

We have European Soccer Database with more than 25000 matches, around 11000 players and other attributes, taken from kaggle website- https://www.kaggle.com/datasets/hugomathien/soccer.


GOALS

1. To extract meaningful insights using SQL Language.
2. To perform Data Analysis useful for futher applications


In [2]:
#IMPORTS
import numpy as np
import sqlite3
import matplotlib.pyplot as plt
import pandas as pd


In [3]:
#path to the database
database= "/content/drive/MyDrive/MLDL/archive (2)/database.sqlite"

In [4]:
#Establish connection and Read all TABLES
conn= sqlite3.connect(database)
tables = pd.read_sql("SELECT * FROM sqlite_master WHERE type='table';", conn)
tables

type               name           tbl_name  rootpage  \
0  table    sqlite_sequence    sqlite_sequence         4   
1  table  Player_Attributes  Player_Attributes        11   
2  table             Player             Player        14   
3  table              Match              Match        18   
4  table             League             League        24   
5  table            Country            Country        26   
6  table               Team               Team        29   
7  table    Team_Attributes    Team_Attributes         2   

                                                 sql  
0             CREATE TABLE sqlite_sequence(name,seq)  
1  CREATE TABLE "Player_Attributes" (\n\t`id`\tIN...  
2  CREATE TABLE `Player` (\n\t`id`\tINTEGER PRIMA...  
3  CREATE TABLE `Match` (\n\t`id`\tINTEGER PRIMAR...  
4  CREATE TABLE `League` (\n\t`id`\tINTEGER PRIMA...  
5  CREATE TABLE `Country` (\n\t`id`\tINTEGER PRIM...  
6  CREATE TABLE "Team" (\n\t`id`\tINTEGER PRIMARY...  
7  CREATE TABLE `Team_Attributes` (\n\t`id`\tINTE...

DISPLAY ALL TABLES FOR BETTER UNDERSTANDIND OF ATTRIBUTES

In [29]:
# TABLE sqlite_sequence
sqlite_sequence= pd.read_sql("select * from sqlite_sequence", conn)
sqlite_sequence

name     seq
0               Team  103916
1            Country   51958
2             League   51958
3              Match   51958
4             Player   11075
5  Player_Attributes  183978
6    Team_Attributes    1458

In [30]:
#TABLE Player_Attributes
Player_Attributes= pd.read_sql("select * from Player_Attributes", conn)
Player_Attributes


id  player_fifa_api_id  player_api_id                 date  \
0            1              218353         505942  2016-02-18 00:00:00   
1            2              218353         505942  2015-11-19 00:00:00   
2            3              218353         505942  2015-09-21 00:00:00   
3            4              218353         505942  2015-03-20 00:00:00   
4            5              218353         505942  2007-02-22 00:00:00   
...        ...                 ...            ...                  ...   
183973  183974              102359          39902  2009-08-30 00:00:00   
183974  183975              102359          39902  2009-02-22 00:00:00   
183975  183976              102359          39902  2008-08-30 00:00:00   
183976  183977              102359          39902  2007-08-30 00:00:00   
183977  183978              102359          39902  2007-02-22 00:00:00   

        overall_rating  potential preferred_foot attacking_work_rate  \
0                 67.0       71.0          right              medium   
1                 67.0       71.0          right              medium   
2                 62.0       66.0          right              medium   
3                 61.0       65.0          right              medium   
4                 61.0       65.0          right              medium   
...                ...        ...            ...                 ...   
183973            83.0       85.0          right              medium   
183974            78.0       80.0          right              medium   
183975            77.0       80.0          right              medium   
183976            78.0       81.0          right              medium   
183977            80.0       81.0          right              medium   

       defensive_work_rate  crossing  ...  vision  penalties  marking  \
0                   medium      49.0  ...    54.0       48.0     65.0   
1                   medium      49.0  ...    54.0       48.0     65.0   
2                   medium      49.0  ...    54.0       48.0     65.0   
3                   medium      48.0  ...    53.0       47.0     62.0   
4                   medium      48.0  ...    53.0       47.0     62.0   
...                    ...       ...  ...     ...        ...      ...   
183973                 low      84.0  ...    88.0       83.0     22.0   
183974                 low      74.0  ...    88.0       70.0     32.0   
183975                 low      74.0  ...    88.0       70.0     32.0   
183976                 low      74.0  ...    88.0       53.0     28.0   
183977                 low      74.0  ...    88.0       53.0     38.0   

        standing_tackle  sliding_tackle  gk_diving  gk_handling  gk_kicking  \
0                  69.0            69.0        6.0         11.0        10.0   
1                  69.0            69.0        6.0         11.0        10.0   
2                  66.0            69.0        6.0         11.0        10.0   
3                  63.0            66.0        5.0         10.0         9.0   
4                  63.0            66.0        5.0         10.0         9.0   
...                 ...             ...        ...          ...         ...   
183973             31.0            30.0        9.0         20.0        84.0   
183974             31.0            30.0        9.0         20.0        73.0   
183975             31.0            30.0        9.0         20.0        73.0   
183976             32.0            30.0        9.0         20.0        73.0   
183977             32.0            30.0        9.0          9.0        78.0   

        gk_positioning  gk_reflexes  
0                  8.0          8.0  
1                  8.0          8.0  
2                  8.0          8.0  
3                  7.0          7.0  
4                  7.0          7.0  
...                ...          ...  
183973            20.0         20.0  
183974            20.0         20.0  
183975            20.0         20.0  
183976            20.0         20.0  
183977        

In [31]:
#TABLE Player
Player= pd.read_sql("select * from Player", conn)
Player


id  player_api_id          player_name  player_fifa_api_id  \
0          1         505942   Aaron Appindangoye              218353   
1          2         155782      Aaron Cresswell              189615   
2          3         162549          Aaron Doran              186170   
3          4          30572        Aaron Galindo              140161   
4          5          23780         Aaron Hughes               17725   
...      ...            ...                  ...                 ...   
11055  11071          26357       Zoumana Camara                2488   
11056  11072         111182         Zsolt Laczko              164680   
11057  11073          36491            Zsolt Low              111191   
11058  11074          35506  Zurab Khizanishvili               47058   
11059  11075          39902   Zvjezdan Misimovic              102359   

                  birthday  height  weight  
0      1992-02-29 00:00:00  182.88     187  
1      1989-12-15 00:00:00  170.18     146  
2      1991-05-13 00:00:00  170.18     163  
3      1982-05-08 00:00:00  182.88     198  
4      1979-11-08 00:00:00  182.88     154  
...                    ...     ...     ...  
11055  1979-04-03 00:00:00  182.88     168  
11056  1986-12-18 00:00:00  182.88     176  
11057  1979-04-29 00:00:00  180.34     154  
11058  1981-10-06 00:00:00  185.42     172  
11059  1982-06-05 00:00:00  180.34     176  

[11060 rows x 7 columns]

In [28]:
#TABLE Match
Match= pd.read_sql("select * from Match", conn)
Match

id  country_id  league_id     season  stage                 date  \
0          1           1          1  2008/2009      1  2008-08-17 00:00:00   
1          2           1          1  2008/2009      1  2008-08-16 00:00:00   
2          3           1          1  2008/2009      1  2008-08-16 00:00:00   
3          4           1          1  2008/2009      1  2008-08-17 00:00:00   
4          5           1          1  2008/2009      1  2008-08-16 00:00:00   
...      ...         ...        ...        ...    ...                  ...   
25974  25975       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25975  25976       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25976  25977       24558      24558  2015/2016      9  2015-09-23 00:00:00   
25977  25978       24558      24558  2015/2016      9  2015-09-22 00:00:00   
25978  25979       24558      24558  2015/2016      9  2015-09-23 00:00:00   

       match_api_id  home_team_api_id  away_team_api_id  home_team_goal  ...  \
0            492473              9987              9993               1  ...   
1            492474             10000              9994               0  ...   
2            492475              9984              8635               0  ...   
3            492476              9991              9998               5  ...   
4            492477              7947              9985               1  ...   
...             ...               ...               ...             ...  ...   
25974       1992091             10190             10191               1  ...   
25975       1992092              9824             10199               1  ...   
25976       1992093              9956             10179               2  ...   
25977       1992094              7896             10243               0  ...   
25978       1992095             10192              9931               4  ...   

        SJA   VCH   VCD   VCA   GBH   GBD   GBA   BSH   BSD   BSA  
0      4.00  1.65  3.40  4.50  1.78  3.25  4.00  1.73  3.40  4.20  
1      3.80  2.00  3.25  3.25  1.85  3.25  3.75  1.91  3.25  3.60  
2      2.50  2.35  3.25  2.65  2.50  3.20  2.50  2.30  3.20  2.75  
3      7.50  1.45  3.75  6.50  1.50  3.75  5.50  1.44  3.75  6.50  
4      1.73  4.50  3.40  1.65  4.50  3.50  1.65  4.75  3.30  1.67  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
25974   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25975   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25976   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25977   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
25978   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[25979 rows x 115 columns]

In [33]:
#TABLE League
League= pd.read_sql("select * from League", conn)
League

id  country_id                      name
0       1           1    Belgium Jupiler League
1    1729        1729    England Premier League
2    4769        4769            France Ligue 1
3    7809        7809     Germany 1. Bundesliga
4   10257       10257             Italy Serie A
5   13274       13274    Netherlands Eredivisie
6   15722       15722        Poland Ekstraklasa
7   17642       17642  Portugal Liga ZON Sagres
8   19694       19694   Scotland Premier League
9   21518       21518           Spain LIGA BBVA
10  24558       24558  Switzerland Super League

In [34]:
#TABLE Country
Country= pd.read_sql("select * from Country", conn)
Country

id         name
0       1      Belgium
1    1729      England
2    4769       France
3    7809      Germany
4   10257        Italy
5   13274  Netherlands
6   15722       Poland
7   17642     Portugal
8   19694     Scotland
9   21518        Spain
10  24558  Switzerland

In [35]:
#TABLE Team
Team	= pd.read_sql("select * from Team	", conn)
Team

id  team_api_id  team_fifa_api_id      team_long_name team_short_name
0        1         9987             673.0            KRC Genk             GEN
1        2         9993             675.0        Beerschot AC             BAC
2        3        10000           15005.0    SV Zulte-Waregem             ZUL
3        4         9994            2007.0    Sporting Lokeren             LOK
4        5         9984            1750.0   KSV Cercle Brugge             CEB
..     ...          ...               ...                 ...             ...
294  49479        10190             898.0       FC St. Gallen             GAL
295  49837        10191            1715.0             FC Thun             THU
296  50201         9777             324.0         Servette FC             SER
297  50204         7730            1862.0  FC Lausanne-Sports             LAU
298  51606         7896               NaN              Lugano             LUG

[299 rows x 5 columns]

In [36]:
#TABLE Team_attributes
Team_attributes	= pd.read_sql("select * from Team_attributes	", conn)
Team_attributes

id  team_fifa_api_id  team_api_id                 date  \
0        1               434         9930  2010-02-22 00:00:00   
1        2               434         9930  2014-09-19 00:00:00   
2        3               434         9930  2015-09-10 00:00:00   
3        4                77         8485  2010-02-22 00:00:00   
4        5                77         8485  2011-02-22 00:00:00   
...    ...               ...          ...                  ...   
1453  1454             15005        10000  2011-02-22 00:00:00   
1454  1455             15005        10000  2012-02-22 00:00:00   
1455  1456             15005        10000  2013-09-20 00:00:00   
1456  1457             15005        10000  2014-09-19 00:00:00   
1457  1458             15005        10000  2015-09-10 00:00:00   

      buildUpPlaySpeed buildUpPlaySpeedClass  buildUpPlayDribbling  \
0                   60              Balanced                   NaN   
1                   52              Balanced                  48.0   
2                   47              Balanced                  41.0   
3                   70                  Fast                   NaN   
4                   47              Balanced                   NaN   
...                ...                   ...                   ...   
1453                52              Balanced                   NaN   
1454                54              Balanced                   NaN   
1455                54              Balanced                   NaN   
1456                54              Balanced                  42.0   
1457                54              Balanced                  42.0   

     buildUpPlayDribblingClass  buildUpPlayPassing buildUpPlayPassingClass  \
0                       Little                  50                   Mixed   
1                       Normal                  56                   Mixed   
2                       Normal                  54                   Mixed   
3                       Little                  70                    Long   
4                       Little                  52                   Mixed   
...                        ...                 ...                     ...   
1453                    Little                  52                   Mixed   
1454                    Little                  51                   Mixed   
1455                    Little                  51                   Mixed   
1456                    Normal                  51                   Mixed   
1457                    Normal                  51                   Mixed   

      ... chanceCreationShooting  chanceCreationShootingClass  \
0     ...                     55                       Normal   
1     ...                     64                       Normal   
2     ...                     64                       Normal   
3     ...                     70                         Lots   
4     ...                     52                       Normal   
...   ...                    ...                          ...   
1453  ...                     53                       Normal   
1454  ...                     50                       Normal   
1455  ...                     32                       Little   
1456  ...                     32                       Little   
1457  ...                     32                       Little   

     chanceCreationPositioningClass  defencePressure defencePressureClass  \
0                         Organised               50               Medium   
1                         Organised               47               Medium   
2                         Organised               47               Medium   
3                         Organised               60               Medium   
4                         Organised               47               Medium   
...                             ...              ...                  ...   
1453                      Organised               46               Medium   
1454                      Organised     

 RAW DATA ANALYSIS

In [5]:
#  TO FIND TOTAL NUMBER OF DISTINCT PLAYERS FROM TABLE PLAYER

distinct_players=pd.read_sql("""select count(distinct (player_api_id) )	from Player """, conn)
distinct_players



count(distinct (player_api_id) )
0                             11060

In [21]:
# TO FIND TOTAL NUMBER OF EACH DISTINCT PLAYERS FROM TABLE PLAYER_ATTRIBUTES

distinct_players1=pd.read_sql("""select distinct(player_api_id) , count(player_api_id) as count
	                             	from Player_Attributes group by player_api_id """, conn)
distinct_players1

player_api_id  count
0               2625     14
1               2752     21
2               2768     21
3               2770     15
4               2790      5
...              ...    ...
11055         744907     11
11056         746419      2
11057         748432      2
11058         750435      9
11059         750584      2

[11060 rows x 2 columns]

So this helps us understand that the total number of distinct players are 11060 and player_api_id is unique id in Player table and a forieng key for Player_Attributes table.

We can also observe that the Match Table is the Fact table with all the "performance" and metrics, but only keys and IDs. While all the descriptive information stored in other tables which are the dimensions.
The reasons is because the DB is designed in a star structure.



DATA INSIGHTS HUNT

1. Let's see the average rating, averge potential and total penalties of each player in the matches played between year 2016-2011.

In [18]:
I1= pd.read_sql(""" select 	Player.player_api_id, Player.player_name , avg(PA.overall_rating) as avg_rating,
                    avg(PA.potential) as avg_potential, sum(PA.penalties) as total_penalties
                    from Player
                    left join Player_Attributes as PA
                    on Player.player_api_id= PA.player_api_id
                    where PA.date between '2011-01-01' and '2016-12-31'
                    group by PA.player_api_id
                    order by avg_rating desc;""", conn)
I1

player_api_id         player_name  avg_rating  avg_potential  \
0              30981        Lionel Messi   93.473684      95.526316   
1              30893   Cristiano Ronaldo   92.000000      93.444444   
2              30955      Andres Iniesta   89.388889      90.777778   
3              30924       Franck Ribery   88.944444      89.388889   
4              35724  Zlatan Ibrahimovic   88.769231      89.076923   
...              ...                 ...         ...            ...   
10376         696435          Jan Bamert   48.000000      67.333333   
10377         528212        Aaron Lennox   48.000000      56.833333   
10378          42149     Florian Berisha   48.000000      62.000000   
10379         674221       Liam Grimshaw   47.714286      60.571429   
10380         614951         Robin Huser   47.000000      61.428571   

       total_penalties  
0                 1462  
1                 1537  
2                 1200  
3                 1463  
4                 1181  
...                ...  
10376              117  
10377              246  
10378               38  
10379              306  
10380              392  

[10381 rows x 5 columns]

This helps us understand the top performers in the last five year from 2011-2016.

2. NOW LET'S FIND OUT THE PERFORMANCE OF EACH TEAM WHO PLAYED FOR SEASON 2015/2016.

In [43]:
a= pd.read_sql(" PRAGMA table_info(Match); ", conn)
a.head(20)

cid              name     type  notnull dflt_value  pk
0     0                id  INTEGER        0       None   1
1     1        country_id  INTEGER        0       None   0
2     2         league_id  INTEGER        0       None   0
3     3            season     TEXT        0       None   0
4     4             stage  INTEGER        0       None   0
5     5              date     TEXT        0       None   0
6     6      match_api_id  INTEGER        0       None   0
7     7  home_team_api_id  INTEGER        0       None   0
8     8  away_team_api_id  INTEGER        0       None   0
9     9    home_team_goal  INTEGER        0       None   0
10   10    away_team_goal  INTEGER        0       None   0
11   11    home_player_X1  INTEGER        0       None   0
12   12    home_player_X2  INTEGER        0       None   0
13   13    home_player_X3  INTEGER        0       None   0
14   14    home_player_X4  INTEGER        0       None   0
15   15    home_player_X5  INTEGER        0       None   0
16   16    home_player_X6  INTEGER        0       None   0
17   17    home_player_X7  INTEGER        0       None   0
18   18    home_player_X8  INTEGER        0       None   0
19   19    home_player_X9  INTEGER        0       None   0

In [72]:
I2=pd.read_sql(""" select match_api_id, team_long_name , count(W.winner) as number_of_wins
                  FROM (select  match_api_id, case
                  when (home_team_goal- away_team_goal) >0 then home_team_api_id
                  when (home_team_goal- away_team_goal)=0 then 0
                  else away_team_api_id
                  end as winner,  T.team_long_name
                  from Match
                  join Team as T
                  on winner= T.team_api_id
                 where season= '2015/2016') AS W
                 group by team_long_name
                 order by number_of_wins desc""", conn)
I2

match_api_id        team_long_name  number_of_wins
0         1989781   Paris Saint-Germain              30
1         2015810            SL Benfica              29
2         2060470              Juventus              29
3         2030085          FC Barcelona              29
4         2030175        Real Madrid CF              28
..            ...                   ...             ...
183       2015917  Académica de Coimbra               5
184       1994757         Polonia Bytom               4
185       1983409            SC Cambuur               3
186       1989993          ES Troyes AC               3
187       1987033           Aston Villa               3

[188 rows x 3 columns]

So we can conclude that in the Season 2015/2016 the team Paris Saint-Germain won 30 matches , hence it is the best performed team of this season.